cmd/dp-dimension-importer/main.go

```diff
@@ -25,61 +41,134 @@ func (r responseBodyReader) Read(reader io.Reader) ([]byte, error) {
-		os.Exit(1)
-	}
+	// Incoming kafka topic for instances to process
+	instanceConsumer := newConsumer(cfg.KafkaAddr, cfg.IncomingInstancesTopic, log.Namespace)
 
-	insertedEventProducer, err := kafka.NewProducer(cfg.KafkaAddr, cfg.DimensionsInsertedTopic, 0)
-	if err != nil {
-		log.ErrorC("kafka producer error", err, log.Data{"topic": cfg.DimensionsInsertedTopic})
-		os.Exit(1)
-	}
+	// Outgoing topic for instances that have completed processing
+	instanceCompleteProducer := newProducer(cfg.KafkaAddr, cfg.OutgoingInstancesTopic)
 
-	var neo4jClient *client.Neo4j
-	if neo4jClient, err = client.NewNeo4j(cfg.DatabaseURL, cfg.PoolSize); err != nil {
-		log.ErrorC("unexpected error while to create database connection pool", err, log.Data{
-			logKeys.URL:      cfg.DatabaseURL,
-			logKeys.PoolSize: cfg.PoolSize,
-		})
-		os.Exit(1)
-	}
+	// Outgoing topic for any errors while processing an instance
+	errorEventProducer := newProducer(cfg.KafkaAddr, cfg.EventReporterTopic)
+
+	neo4jClient := createNeo4jClient(cfg)
 
 	newDimensionInserterFunc := func() handler.DimensionRepository {
 		return repository.NewDimensionRepository(neo4jClient, map[string]string{})
 	}
 
-	datasetAPI := client.DatasetAPI{
-		ResponseBodyReader:  responseBodyReader{},
-		HTTPClient:          &http.Client{},
+	// ImportAPI HTTP client.
+	datasetAPICli := client.DatasetAPI{
 		DatasetAPIHost:      cfg.DatasetAPIAddr,
 		DatasetAPIAuthToken: cfg.DatasetAPIAuthToken,
+		HTTPClient:          &http.Client{},
+		ResponseBodyReader:  responseBodyReader{},
 	}
 
-	eventHandler := &handler.DimensionsExtractedEventHandler{
+	// Handler for dimensionsExtracted events.
+	instanceEventHandler := &handler.InstanceEventHandler{
 		NewDimensionInserter: newDimensionInserterFunc,
+		DatasetAPICli:        datasetAPICli,
 		InstanceRepository:   &repository.InstanceRepository{Neo4j: neo4jClient},
-		DatasetAPI:           datasetAPI,
 	}
 
-	dimensionInsertedProducer := message.DimensionInsertedProducer{
-		Producer:   insertedEventProducer,
-		Marshaller: schema.DimensionsInsertedSchema,
+	// MessageProducer for dimensionsInsertedEvents
+	instanceCompletedProducer := message.InstanceCompletedProducer{
+		Producer:   instanceCompleteProducer,
+		Marshaller: schema.InstanceCompletedSchema,
+	}
+
+	// Errors handler
+	errorEventHandler := &handler.ErrorHandler{
+		Producer:   errorEventProducer,
+		Marshaller: schema.ErrorEventSchema,
 	}
 
-	exitChannel := make(chan error)
-	err = message.Consume(extractedEventConsumer, dimensionInsertedProducer, eventHandler, exitChannel)
+	healthCheckErrors := make(chan error)
+
+	// HTTP Health check endpoint.
+	healthcheck.NewHandler(cfg.BindAddr, healthCheckErrors)
+
+	ctx, cancelConsumerLoop := context.WithCancel(context.Background())
+
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
+		log.Info(gracefulShutdownMsg, nil)
+		shutdownCTX, _ := context.WithTimeout(ctx, cfg.ShutdownTimeout)
```

Comment:
Should be handling `cancel` function returned by `context.WithTimeout()`, like so:

```
shutdownCTX, cancel := context.WithTimeout(ctx, cfg.ShutdownTimeout)
defer cancel()
```